keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,360 @@
+
+### Approach C - Use conditions
+
+This design continues the existing pattern of expressing health of various node features and components via Conditions.
+For example the Kubelet uses conditions heavily and the Node Problem Detector also uses conditions. Conditions allow for recording heartbeats already.
+
+This design is similar to the approach `A` where instead of recording state to a separate object, kubelet and other system components will record their state via Conditions.
+The node controller will be extended to support a configurable policy on what Conditions influence the meta readiness of a node.
+
+The kubelet and node controller will use a taint similar to the one illustrated in approach `A` to signal node meta readiness.
+
+The scheduler needs to understand that certain special pods (system pods) can be scheduled to nodes even if the node isn’t logically ready.
+There is no canonical construct today to differentiate system pods from regular pods.
+
+The main drawback of this approach is the explosion on heartbeats (updates to Condition objects) to the Node object from various components which was the reason why a new heartbeat mechanism was introduced for the kubelet in the first place.
+Another drawback is that Conditions are not extensible - it is not possible to include extension specific information as part of conditions leading to leaking the state of extensions to other sub-objects in the Node API object eventually.
+
+### Approach D - Support extensible health checks in the Kubelet
+
+This approach assumes kubelet to be the central arbitrator of Node health.
+
+Every node extension needs to register a canonical Readiness endpoint (similar to pod readiness/handlers) with the kubelet via the Node API object.
+
+The kubelet will then poll health of individual extensions periodically and reflect the state of the individual extensions via a taint similar to the solutions mentioned above.
+The main difference here is that the node controller doesn’t have to track the state of individual extensions since the kubelet is already doing it.
+
+Optionally, the kubelet can reflect the state of each extension via a dedicated condition.
+
+Upon restart, Kubelet will wait for one successful health check run across all registered extensions prior to removing the special taint.
+
+This approach requires that all extensions are accessible to the kubelet via host network interfaces (barring the option of using exec binaries per addon that will suffer a distribution problem).
+
+This approach also lacks the ability to include more state information about each extension in the API object - aka it’s limited to only health tracking. The load on the kubelet could be much higher due to the proliferation of system extensions. Every extension is required to build a canonical healthz endpoint (which is a best practice by itself).
+
+### Risks and Mitigations
+
+TODO
+
+## Design Details
```

Comment:
We also need to include a Production Maintenance section that talks about the health metrics and debugging tips to safely rollout and manage this feature in production.